signer/signer.go

```diff
@@ -0,0 +1,483 @@
+	Value string `asn1:"optional,ia5"`
+}
+
+type policyInformation struct {
+	Policy     asn1.ObjectIdentifier
+	Qualifiers []policyQualifier `asn1:"optional"`
+}
+
+var stringToQualifierType = map[string]asn1.ObjectIdentifier{
+	"id-qt-cps": {1, 3, 6, 1, 5, 5, 7, 2, 1},
+}
+
+func newProfile(config ProfileConfig) (*signingProfile, error) {
+	sp := &signingProfile{
+		allowRSAKeys:    config.AllowRSAKeys,
+		allowECDSAKeys:  config.AllowECDSAKeys,
+		allowMustStaple: config.AllowMustStaple,
+		allowCTPoison:   config.AllowCTPoison,
+		allowSCTList:    config.AllowSCTList,
+		allowCommonName: config.AllowCommonName,
+		issuerURL:       config.IssuerURL,
+		crlURL:          config.CRLURL,
+		ocspURL:         config.OCSPURL,
+		maxBackdate:     config.MaxValidityBackdate.Duration,
+		maxValidity:     config.MaxValidityPeriod.Duration,
+	}
+	if config.IssuerURL == "" {
+		return nil, errors.New("Issuer URL is required")
+	}
+	if config.OCSPURL == "" {
+		return nil, errors.New("OCSP URL is required")
+	}
+	if len(config.Policies) > 0 {
+		var policies []policyInformation
+		for _, policyConfig := range config.Policies {
+			id, err := parseOID(policyConfig.OID)
+			if err != nil {
+				return nil, fmt.Errorf("failed parsing policy OID %q: %s", policyConfig.OID, err)
+			}
+			pi := policyInformation{Policy: id}
+			for _, qualifierConfig := range policyConfig.Qualifiers {
+				qt, ok := stringToQualifierType[qualifierConfig.Type]
+				if !ok {
+					return nil, fmt.Errorf("unknown qualifier type: %s", qualifierConfig.Type)
+				}
+				pq := policyQualifier{
+					OID:   qt,
+					Value: qualifierConfig.Value,
+				}
+				pi.Qualifiers = append(pi.Qualifiers, pq)
+			}
+			policies = append(policies, pi)
+		}
+		policyExtBytes, err := asn1.Marshal(policies)
+		if err != nil {
+			return nil, err
+		}
+		sp.policies = &pkix.Extension{
+			Id:    asn1.ObjectIdentifier{2, 5, 29, 32},
+			Value: policyExtBytes,
+		}
+	}
+	return sp, nil
+}
+
+// requestValid verifies the passed IssuanceRequest against the signingProfile. If the
+// request doesn't match the signing profile an error is returned.
+func (p *signingProfile) requestValid(clk clock.Clock, req *IssuanceRequest) error {
+	switch req.PublicKey.(type) {
+	case *rsa.PublicKey:
+		if !p.allowRSAKeys {
+			return errors.New("RSA keys not allowed")
+		}
+	case *ecdsa.PublicKey:
+		if !p.allowECDSAKeys {
+			return errors.New("ECDSA keys not allowed")
+		}
+	default:
+		return errors.New("unsupported public key type")
+	}
+
+	if !p.allowMustStaple && req.IncludeMustStaple {
+		return errors.New("must-staple extension cannot be included")
+	}
+
+	if !p.allowCTPoison && req.IncludeCTPoison {
+		return errors.New("ct poison extension cannot be included")
+	}
+
+	if !p.allowSCTList && req.SCTList != nil {
+		return errors.New("sct list extension cannot be included")
+	}
+
+	if req.IncludeCTPoison && req.SCTList != nil {
+		return errors.New("cannot include both ct poison and sct list extensions")
+	}
+
```

Comment:
I think @orangepizza raises a good point that `if !req.IncludeCTPoison && req.SCTList == nil` (i.e. the request doesn't have SCTs _and also_ doesn't have the precert poison extension) should result in rejection here too?

Reply:
This was discussed OOB, this is not really a technical requirement, and starts bleeding policy into the signer, which is already enforced elsewhere. There weren't incredibly strong opinions on either side of leaving vs. changing, so I'm just going to leave as-is.